Generalized Strang splitting

jagot opened the pull request:
I made some simple changes to allow for `AffineDiffEqOperator`:s composed of an arbitrary number of `AbstractDiffEqOperator`:s; this implements #116. Per [discussions on Gitter](https://gitter.im/JuliaDiffEq/Lobby?at=5992c74bee5c9a4c5fc0e877), I renamed `StrangSplitting` to `MidpointSplitting`, since that is what we are implementing here; Strang splitting is more general than this. In keeping with this, the RHS is now evaluated at the midpoint of the interval.

In `perform_step!`, I use a small `@swap!` macro, I am unsure if there is a better place to put this.

As of now, only out-of-place `expm` is supported (only suitable for small systems), but I submit this PR to see if I am going in the right direction, before starting on #115.

ChrisRackauckas commented on test/linear_method_tests.jl:
```diff
@@ -30,5 +30,37 @@ function (p::typeof(L))(::Type{Val{:analytic}},t,u0)
 prob = ODEProblem(L,u0,(0.0,1.0))
-sol = solve(prob,StrangSplitting(),dt=1/10)
+sol = solve(prob,MidpointSplitting(),dt=1/10)
 using Plots; pyplot; plot(sol)
```
Can't plot in tests because it'll bog down the CI which is why no plotting package is a dependency. This'll need to be commented out.

ChrisRackauckas commented:
Can you add a convergence test? Example:

```julia
dts = 1./2.^(7:-1:4)
sim  = test_convergence(dts,prob,MidpointSplitting())
@test abs(sim.𝒪est[:l2]-2) < 0.2
```

If the analytic solution is set this'll be all that's needed. You can check it via `plot(sim)`

jagot commented:
Sorry for the delay, I have implemented a simple convergence test and rebased onto master. The convergence test is trivially generalizable to the complex case, as soon as JuliaDiffEq/DiffEqDevTools.jl#4 is fixed.

ChrisRackauckas commented on src/integrators/linear_integrators.jl:
```diff
@@ -118,20 +118,37 @@ function initialize!(integrator,cache::StrangSplittingCache,f=integrator.f)
 end
 
-function perform_step!(integrator,cache::StrangSplittingCache,f=integrator.f)
+macro swap!(x,y)
```
This should go to utils

ChrisRackauckas commented:
Thanks! Note that this won't be part of the normal testing until DiffEqOperators.jl is registered, which I hope is sooner rather than later.
